compiler/dialects/kernel.py

```diff
@@ -0,0 +1,58 @@
+from abc import ABC
+
+from xdsl.dialects.builtin import IntegerType
+from xdsl.ir import Dialect
+from xdsl.irdl import operand_def
+from xdsl.irdl.operations import irdl_op_definition, result_def
+from xdsl.parser import IRDLOperation
+
+
+class KernelOp(IRDLOperation, ABC):
+    ...
+
+
+class BinaryOp:
+    lhs = operand_def(IntegerType)
+    rhs = operand_def(IntegerType)
+    result = result_def(IntegerType)
+
+
+class QuantizedBinaryOp(BinaryOp):
+    zp_lhs = operand_def(IntegerType)
+    zp_rhs = operand_def(IntegerType)
+
+
+@irdl_op_definition
+class MulOp(KernelOp, BinaryOp):
+    name = "kernel.mul"
+    assembly_format = (
+        "$lhs `,` $rhs attr-dict `:` type($lhs) `,` type($rhs) `->` type($result)"
+    )
+
+
+@irdl_op_definition
+class AddOp(KernelOp, BinaryOp):
+    name = "kernel.add"
+    assembly_format = (
+        "$lhs `,` $rhs attr-dict `:` type($lhs) `,` type($rhs) `->` type($result)"
+    )
+
+
+@irdl_op_definition
+class MacOp(KernelOp, BinaryOp):
+    name = "kernel.mac"
+    assembly_format = (
+        "$lhs `,` $rhs attr-dict `:`" "type($lhs) `,` type($rhs) `->` type($result)"
+    )
```

Comment:
Why is this slightly different from the others?
Also, can we not move the assembly format to either ABCs?

Reply:
apperently moving them into the abc's breaks xdsl, and it is not recognized. something to do with the @irdl_op_definition magic